packages/loopring_v3/contracts/impl/BasicDepositContract.sol

```diff
@@ -110,16 +94,27 @@ contract BasicDepositContract is IDepositContract, ReentrancyGuard
                 );
             }
         }
+
+        uint exchangeBalanceBefore = exchangeBalance[token];
+        // Keep track how many tokens are deposited in the exchange
+        uint exchangeBalanceAfter = exchangeBalanceBefore.add(amount);
+        // Make sure the total max amount per token in the exchange is capped
+        require(exchangeBalanceAfter <= MAX_TOTAL_TOKEN_BALANCE, "MAX_AMOUNT_REACHED");
```

Comment:
I think we only need to make sure the deposit amount is less than this value, not the total accumulated balance.

The relayer will also need to check the accumulated balance of each user in each token is less than 2 ** 96 - 1;

Reply:
It's the total balance that needs to be be < `MAX_TOTAL_TOKEN_BALANCE`, at least now. This is actually pretty important because if the balance for a single users is already `MAX_TOTAL_TOKEN_BALANCE`, we would need to cap the updated balance when there is interest. And probably other things like that, so there are certainly a couple of edge cases that would introduce. Easiest is just to keep the total balance < `MAX_TOTAL_TOKEN_BALANCE`, no need to worry about overflow anywhere in the circuits and should be more than enough for any token.

Reply:
Lets clarify a few concepts:
- `exchangeBalanceAfter`: this is the total amount of token deposited by all users after this new deposit, only visible to the DepositCongtract.
- user's token balance: this is the layer-2's notation of a user's total balance for a given token. 
- deposit amount: this is the pending total amount of token for a user to be deposited.

I think the last two should always be smaller than  MAX_TOTAL_TOKEN_BALANCE, but not `exchangeBalanceAfter`. As the circuit doesn't even know of this variable at all. Correct?

Reply:
> exchangeBalanceAfter: this is the total amount of token deposited by all users after this new deposit, only visible to the DepositCongtract.

It'll be the total balance after profits. So it will be at least as large as the total sum of all deposited/withdrawn balances.

> I think the last two should always be smaller than MAX_TOTAL_TOKEN_BALANCE, but not exchangeBalanceAfter. As the circuit doesn't even know of this variable at all. Correct?

In theory yes, but if a single user has ` MAX_TOTAL_TOKEN_BALANCE` we need the cap the amount if he gains any interest on it. This (and maybe some other cases) would be special cases that the circuit would need to handle if we only force ` MAX_TOTAL_TOKEN_BALANCE` for each balance instead of the total rollup balance. I don't think this is really necessary so I prefer to do this single check onchain instead of adding complexity on the circuit logic to handle these special cases.